contracts/factory/src/contract.rs

```diff
@@ -185,54 +187,78 @@ pub fn instantiate(
-    let rewards_manager_contract = deps
-        .api
-        .addr_humanize(&rewards_manager_address)?
-        .to_string();
-    let rewards_pump_contract = deps.api.addr_humanize(&rewards_pump_address)?.to_string();
-    let splitter_contract = deps.api.addr_humanize(&splitter_address)?.to_string();
-    let lsm_share_bond_provider_contract = deps
-        .api
-        .addr_humanize(&lsm_share_bond_provider_address)?
-        .to_string();
-    let native_bond_provider_contract = deps
-        .api
-        .addr_humanize(&native_bond_provider_address)?
-        .to_string();
-
-    let state = State {
-        token_contract: token_contract.to_string(),
-        core_contract: core_contract.to_string(),
-        puppeteer_contract: puppeteer_contract.to_string(),
-        withdrawal_voucher_contract: withdrawal_voucher_contract.to_string(),
-        withdrawal_manager_contract: withdrawal_manager_contract.to_string(),
-        strategy_contract: strategy_contract.to_string(),
-        validators_set_contract: validators_set_contract.to_string(),
-        distribution_contract: distribution_contract.to_string(),
-        rewards_manager_contract: rewards_manager_contract.to_string(),
-        rewards_pump_contract: rewards_pump_contract.to_string(),
-        splitter_contract: splitter_contract.to_string(),
-        lsm_share_bond_provider_contract: lsm_share_bond_provider_contract.to_string(),
-        native_bond_provider_contract: native_bond_provider_contract.to_string(),
-    };
-    STATE.save(deps.storage, &state)?;
+    let core_contract = deps.api.addr_humanize(&core_address)?;
+    let token_contract = deps.api.addr_humanize(&token_address)?;
+    let withdrawal_voucher_contract = deps.api.addr_humanize(&withdrawal_voucher_address)?;
+    let withdrawal_manager_contract = deps.api.addr_humanize(&withdrawal_manager_address)?;
+    let strategy_contract = deps.api.addr_humanize(&strategy_address)?;
+    let validators_set_contract = deps.api.addr_humanize(&validators_set_address)?;
+    let distribution_contract = deps.api.addr_humanize(&distribution_calculator_address)?;
+    let puppeteer_contract = deps.api.addr_humanize(&puppeteer_address)?;
+    let rewards_manager_contract = deps.api.addr_humanize(&rewards_manager_address)?;
+    let rewards_pump_contract = deps.api.addr_humanize(&rewards_pump_address)?;
+    let splitter_contract = deps.api.addr_humanize(&splitter_address)?;
+    let lsm_share_bond_provider_contract =
+        deps.api.addr_humanize(&lsm_share_bond_provider_address)?;
+    let native_bond_provider_contract = deps.api.addr_humanize(&native_bond_provider_address)?;
+
+    STATE.save(deps.storage, "token_contract".to_string(), &token_contract)?;
+    STATE.save(deps.storage, "core_contract".to_string(), &core_contract)?;
+    STATE.save(
+        deps.storage,
+        "puppeteer_contract".to_string(),
+        &puppeteer_contract,
+    )?;
```

Comment:
Maybe this state should really be a map? I am somewhat afraid of using unprefixed storage keys

Reply:
When I was talking about a map, I was meaning `cw_storage_plus::Map` and not `std::collections::HashMap` :thinking: 

In current implementation, to read a single address from storage, first one has to load the entire map into memory.